### PR TITLE
Unify IR operator representation (stop using attributes in the JIT)

### PR DESCRIPTION
**Pull request by apaszke**
Based on top of #9763 (first 3 commits belong to that PR). The first commits from this PR are "Stop using attributes ..."

I tried to separate the changes into fairly meaningful commits. I can't split them up into smaller PRs, because everything starts working and all tests pass only after the whole sequence, but hopefully this will make reviewing somewhat easier.

Known issues/regressions/future tasks:
- `aten::lerp` and `aten::clamp` are no longer fusable
- `CreateAutodiffSubgraphs` needs a rewrite
  - It is much more strict now, and will miss a lot of opportunities, especially when viewing ops are involved. Our previous approach was "ignore the assumption on shape availability in gradient formulas to determine differentiability, and hope that shape prop will be robust enough to actually deliver them before we differentiate", which obviously doesn't scale well to more complex cases. We should either work on reducing the size dependency of grad formulas (feasible e.g. for `view`/`reshape`, unfeasible for `squeeze`/`unsqueeze`), or make `CreateAutodiffSubgraphs` integrate some kind of "I could integrate this node into an AD subgraph, but will I be able to infer the shape of its input" reasoning (kind of like a limited shape prop, that doesn't infer anything, and only tells if it *could* infer something).
  - It sometimes creates constant-only (or constants + one node) graphs, which is useless
- Broken `aten::add` in auto-batching, because it gained a non-tensor input. I changed the test for pointwise operations to use `aten::mul` instead, but I needed to disable the LSTM cell test. I'm not sure how scalar constants should be implemented in this case, because I don't fully understand our format. cc: @ChunliF 
- Graph import does some hacks to recover type of constants. This code should be removed once we'll gain the ability to export the IR along with value types.
- There's still a fair amount of dead code that can be removed. I didn't want to make this diff any bigger, and removing it is an easy task.
- Graph fuser could be improved to use signature matching (possibly using `OperatorSet`) instead of basing on node kinds.
- Manual constant propagation for the `ListConstruct` node in `torch/onnx/utils.py` should be replaced with a proper constant propagation pass (or we should ensure that the one we have handles at least this case before we remove this code).

@zdevito

### CODE REVIEW

**Review by zdevito — APPROVED** [review]
This looks good. I only had minor comments. I like the approach to fixing the ONNX symbolics. It keeps them working without burdening us with expanding them to non-const inputs yet.

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -387,55 +381,22 @@ def emit_record_trace(env):
...
-            local['record_positional_attributes'].append(
-                RECORD_POSITIONAL_ATTRIBUTE.substitute(name=arg['name'], i=i))
+                return 'trace_inputs.push_back({});'.format(arg['name'])
+            elif arg['simple_type'] == 'TensorList':
+                return 'trace_inputs.insert(trace_inputs.end(), {n}.begin(), {n}.end());'.format(n=arg['name'])
+            elif arg['simple_type'] == 'IntList':
+                return 'trace_inputs.push_back(make_variable(getIntListTensor("{n}", {n}), false));'.format(
+                    n=arg['name'])
+            elif arg['simple_type'] in {'bool', 'double', 'int64_t', 'Scalar'}:
+                return 'trace_inputs.push_back(make_variable(jit::as_tensor({}), false));'.format(arg['name'])
```
[the remark this answers is not in the document]
Comment (apaszke) [reply tools/autograd/gen_variable_type.py]: Looks like it's dead (it was used to stop emitting attributes before we merged the thread-local tracer). Will remove.

File: torch/csrc/jit/autodiff.cpp [hunk 2]
```diff
@@ -83,146 +116,149 @@ bool outputRequiresGrad(Node* node, std::function<bool(Value*)> requires_grad) {
...
-        SymbolicVariable dmat1, dmat2;
-        if (auto type = inputs.at(0).value()->type()->cast<TensorType>()) {
-          auto sizes = type->sizes(), strides = type->strides();
-          if (strides.at(0) == 1 && strides.at(1) == sizes.at(0)) {
-            dmat1 = inputs.at(1).mm(grads.at(0).t()).t();
-          } else {
-            dmat1 = grads.at(0).mm(inputs.at(1).t());
-          }
+
+    if (node->matches("aten::add(Tensor self, Tensor other, *, Scalar alpha) -> Tensor") ||
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/autodiff.cpp]: Oh they definitely are. And they at least give you enough confidence to safely access and process input lists to nodes. I'm only looking forward to getting rid of flattening tensor lists into arguments, and we'll be in a very good place then.

File: torch/onnx/symbolic.py [hunk 3]
```diff
@@ -208,22 +269,25 @@ def sigmoid(g, self):
     return g.op("Sigmoid", self)
 
 
+# TODO (apaszke): FIX FIX FIX FIX
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/onnx/symbolic.py]: I meant to and I did 😊Will remove the comment.

File: torch/csrc/jit/autodiff.cpp [hunk 4]
```diff
@@ -13,40 +15,71 @@ namespace torch { namespace jit {
...
+    return n->is_constant(attr::size) && n->is_constant(attr::implicit);
+  }
+  if (n->matches("aten::view(Tensor self, int[] size) -> Tensor") ||
+      n->matches("aten::reshape(Tensor self, int[] shape) -> Tensor")) {
+    return n->namedInput(attr::self)->type()->cast<TensorType>();
   }
 
   // linear blocks may appear as inputs to graph executors, but they are removed
   // before differentiation occurs
+  // TODO (apaszke): is this really needed?
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/jit/autodiff.cpp]: I just don't understand under what conditions we would be checking differentiability of the `GradOf` blocks. Shouldn't their lowering be one of the first passes that we would possibly run? It should happen before `CreateAutodiffSubgraphs` and `Differentiate` for sure, right?

File: torch/onnx/symbolic.py [hunk 5]
```diff
@@ -631,57 +712,55 @@ def abs(g, self):
 
 
 def pow(g, self, exponent):
+    exponent = _maybe_get_scalar(exponent)
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/onnx/symbolic.py]: They do support scalars. The point of this `_maybe_get_scalar` function is to only return a constant tensor if it's also a scalar (if it's a constant, but not a scalar, we want to use the Value)

File: torch/onnx/symbolic.py [hunk 6]
```diff
@@ -631,57 +712,55 @@ def abs(g, self):
...
 
 
-def min(g, self, *args, **kwargs):
-    dim = kwargs.get("dim", None)
-    if dim is None and isinstance(args[0], numbers.Number):
-        dim = args[0]
-    if dim is not None:
-        keepdim = kwargs.get("keepdim", False)
-        # TODO: export it as ReduceMin
+def min(g, self, dim_or_y, keepdim=None):
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/onnx/symbolic.py]: Because we want to work with raw values. `parse_args` is there only to automatically convert Values into constants (so that we don't have to change most of the symbolics, which previously assumed those would be delivered as constants via attributes)

File: torch/onnx/symbolic.py [hunk 7]
```diff
@@ -32,6 +32,59 @@
 # ---------------------------------------------------------------------
 
 
+def _parse_arg(value, desc):
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/onnx/symbolic.py]: Will add them in a next patch that will be less likely to conflict.
